src/darwin/Framework/CHIPTests/MTROTAProviderTests.m

```diff
@@ -864,7 +898,437 @@ - (void)test005_DoBDXTransferAllowUpdateRequest
+            [readHandle seekToOffset:offset error:&seekError];
+            XCTAssertNil(seekError);
+
+            NSError * readError = nil;
+            NSData * data = [readHandle readDataUpToLength:blockSize.unsignedLongValue error:&readError];
+            XCTAssertNil(readError);
+            XCTAssertNotNil(data);
+
+            BOOL isEOF = offset + blockSize.unsignedLongValue >= imageSize;
+
+            ++lastBlockIndex;
+
+            if (isEOF) {
+                sOTAProviderDelegate.blockQueryHandler = nil;
+            }
+
+            completion(data, isEOF);
+
+            if (isEOF) {
+                [bdxQueryExpectation2 fulfill];
+            }
+        };
+        sOTAProviderDelegate.transferEndHandler
+            = ^(NSNumber * nodeID, MTRDeviceController * controller, NSError * _Nullable error) {
+                  XCTAssertEqualObjects(nodeID, @(kDeviceId2));
+                  XCTAssertEqual(controller, sController);
+                  XCTAssertNil(error);
+                  sOTAProviderDelegate.transferEndHandler = nil;
+                  [bdxEndExpectation2 fulfill];
+              };
+    };
+
+    // Set up the apply update request and update applied handlers for device1
+    sOTAProviderDelegate.applyUpdateRequestHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+        MTROTASoftwareUpdateProviderClusterApplyUpdateRequestParams * params, ApplyUpdateRequestCompletion completion) {
+        XCTAssertEqualObjects(nodeID, @(kDeviceId1));
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(params.updateToken, updateToken);
+        XCTAssertEqualObjects(params.newVersion, kUpdatedSoftwareVersion2); // TODO: Factor this out better!
+
+        XCTAssertTrue([[NSFileManager defaultManager] contentsEqualAtPath:otaRawImagePath1 andPath:kOtaDownloadedFilePath1]);
+
+        sOTAProviderDelegate.applyUpdateRequestHandler = nil;
+        [sOTAProviderDelegate respondToApplyUpdateRequestWithAction:MTROTASoftwareUpdateProviderOTAApplyUpdateActionProceed
+                                                         completion:completion];
+        [applyUpdateRequestExpectation1 fulfill];
+    };
+
+    sOTAProviderDelegate.notifyUpdateAppliedHandler = ^(NSNumber * nodeID, MTRDeviceController * controller,
+        MTROTASoftwareUpdateProviderClusterNotifyUpdateAppliedParams * params, MTRStatusCompletion completion) {
+        XCTAssertEqualObjects(nodeID, @(kDeviceId1));
+        XCTAssertEqual(controller, sController);
+        XCTAssertEqualObjects(params.updateToken, updateToken);
+        XCTAssertEqualObjects(params.softwareVersion, kUpdatedSoftwareVersion2);
+        sOTAProviderDelegate.notifyUpdateAppliedHandler = nil;
+        [sOTAProviderDelegate respondSuccess:completion];
+        [notifyUpdateAppliedExpectation1 fulfill];
+
+        // Update has been applied for device1
+        // Set up the apply update request and update applied handlers for device2
```

Comment:
device2 could end up applying its update before device1 does.  It could even apply its update before device1 gets to applyUpdateRequestHandler.  Nothing enforces the ordering assumed here.

We need to either not respond with "image available" until device1 has applied its update and notified us, or if that would not fit in with what this test is trying to test we need to accept that the update request and update applied bits can come for the two devices in any order (and for example check the value of nodeID in the handlers to decide what tests to do and which expectations to fulfill).

Reply:
done.